python_modules/libraries/dagster-databricks/dagster_databricks/databricks_step_main.py

```diff
@@ -0,0 +1,121 @@
+"""
+The main entrypoint for solids executed using the DatabricksPySparkStepLauncher.
+
+This script is launched on Databricks using a `spark_python_task` and is passed the following
+parameters:
+
+- the DBFS path to the pickled `step_run_ref` file
+- the DBFS path to the zipped pipeline
+- paths to any other zipped packages which have been uploaded to DBFS.
+"""
+
+import os
+import pickle
+import site
+import sys
+import tempfile
+import zipfile
+
+from dagster.core.execution.plan.external_step import PICKLED_EVENTS_FILE_NAME, run_step_from_ref
+from dagster.serdes import serialize_value
+
+
+# This won't be set in Databricks but is needed to be non-None for the
+# Dagster step to run.
+if 'DATABRICKS_TOKEN' not in os.environ:
+    os.environ['DATABRICKS_TOKEN'] = ''
+
+
+def setup_s3_storage(scope, credentials):
+    """
+    Obtain AWS credentials from Databricks secrets and export so both Spark and boto can use them.
+    """
+    access_key = dbutils.secrets.get(scope=scope, key=credentials['access_key_key'])
+    secret_key = dbutils.secrets.get(scope=scope, key=credentials['secret_key_key'])
+
+    # Spark APIs will use this.
+    sc._jsc.hadoopConfiguration().set('fs.s3n.awsAccessKeyId', access_key)
+    sc._jsc.hadoopConfiguration().set('fs.s3n.awsSecretAccessKey', secret_key)
+
+    # Boto will use these.
+    os.environ['AWS_ACCESS_KEY_ID'] = access_key
+    os.environ['AWS_SECRET_ACCESS_KEY'] = secret_key
+
+
+def setup_adls2_storage(scope, credentials):
+    """
+    Obtain an Azure Storage Account key from Databricks secrets and export so Spark can use it.
+    """
+    storage_account_key = dbutils.secrets.get(
+        scope=scope, key=credentials['storage_account_key_key']
+    )
+    sc._jsc.hadoopConfiguration().set(
+        'fs.azure.account.key.{}.dfs.core.windows.net'.format(credentials['storage_account_name']),
+        storage_account_key,
+    )
+
+
+def setup_storage(step_run_ref):
```

Comment:
Currently, the preferred way in dagster to configure a storage like this is to set it up as an IntermediateStore - have you looked into that?

Reply:
The problem I found was that this `main` script is executed on Databricks so the environment is different (e.g. boto credentials probably aren't available) and Spark needs to be set up in a special way to be able to access S3/ADLS2.

In other words, the intermediate storage is still used, but in the Databricks context we need to convert Databricks Secrets to something Boto/Spark can understand before trying to access it. E.g. `S3IntermediateStore` assumes (via `S3ObjectStore`) that boto will be configured, so we need to set the env vars; `SparkDataFrameS3StoragePlugin` assumes that the `spark_session` of the `pyspark` resource will be able to read/write parquet files using paths, so we need to set up the spark session (I've linked to the Databricks documentation about accessing storage, now).

I may be completely missing the point though - I agree that it's a bit messy to have to do it here! :smile:

Reply:
Got it, now I see why this is tough.  I'm interested in your thoughts on an alternative approach: put this special handling inside the `SparkDataFrameS3StoragePlugin`.

This would require some way of detecting from within the `SparkDataFrameS3StoragePlugin` that we're running on databricks.  I'm guessing there's some sort of environment variable that we could check if exists?  Or could set a flag from databricks_step_main?

The advantages would be:
* A little more consistent with how we've handled similar issues in the past (e.g. using the "s3" instead of "s3a" protocol when running in EMR).
* Puts this storage-specific logic in a place that's already scoped to a particular storage.

Reply:
CC @alangenfeld who I was chatting about this with.

Reply:
> Got it, now I see why this is tough. I'm interested in your thoughts on an alternative approach: put this special handling inside the `SparkDataFrameS3StoragePlugin`.
> 
> This would require some way of detecting from within the `SparkDataFrameS3StoragePlugin` that we're running on databricks. I'm guessing there's some sort of environment variable that we could check if exists? Or could set a flag from databricks_step_main?
> 
> The advantages would be:
> 
>     * A little more consistent with how we've handled similar issues in the past (e.g. using the "s3" instead of "s3a" protocol when running in EMR).
> 
>     * Puts this storage-specific logic in a place that's already scoped to a particular storage.

This definitely sounds like the right idea for DataFrames specifically.

I noticed that I also had to set up AWS credentials for boto otherwise complained that it had no valid creds. Presumably that's because the S3 storage system assumes it's configured with boto creds. If that's the case, then the `SparkDataFrameS3StoragePlugin` doesn't feel like the right place to set up those environment variables for system storage, but I don't know where the right place is...

Reply:
Ah, that's a good point.  I've racked my brain and can't really think of a good way of dealing with that, without adding some new abstractions to the guts of Dagster or causing a circular dependency between dagster_aws and dagster_databricks.

Given that, I think it's probably best to keep things consistent and leave the DataFrame stuff in here, instead of implementing my suggestion to put it in SparkDataFrameS3StoragePlugin.

I've filed an issue to track cleaning this up in the future: https://github.com/dagster-io/dagster/issues/2492.  Mind referencing it here in a comment?

Reply:
I've added a link to that issue in databricks_step_main.py :+1: